### ISSUE TITLE
Kilosort 3.0 detecting too few clusters on 4-shank data

### ISSUE THREAD

**Issue by msj220001** [MouseLand/Kilosort#657]
### Describe the issue:

I'm using Kilosort 3.0 to do spike sorting on data from a Neuronexus A4x4 CM16LP probe.

With the parameters I have right now, Kilosort is only detecting 8 units and only 4 have clear units.

From manually sorting this file, there should be 32 units (on average 2 per channel).

![image](https://github.com/MouseLand/Kilosort/assets/157846284/f8e4b697-d292-41e0-9ef0-9b90fe80018f)

Here are my current parameters:
![image](https://github.com/MouseLand/Kilosort/assets/157846284/a867e31a-b75d-4929-b342-f0984183e8d0)

Which parameters should I adjust if I want kilosort to identify more units and clusters?

This is a separate question: but phy is displaying multiple channels' waveforms in the Waveform View, but my kcoords are [1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16] which should make kilosort identify units on an individual/isolated channel basis, correct?

**Comment by marius10p** [comment]
Can you please try Kilosort4?

**Comment by msj220001** [comment]
Yes, I've been running Kilosort 4 as well and I am a bit confused by some of the parameters, specifically the threshold parameters.

Correct me if I'm wrong, does Th(universal) set the 'lower bound' in standard deviations for waveforms selected for spike detection and Th(learned) sets the 'upper bound' that if a waveform exceeds it is considered an artifact? Is that the meaning of these parameters?
![Screenshot 2024-04-16 165718](https://github.com/MouseLand/Kilosort/assets/157846284/b0a9fe91-5bc2-4633-af2f-22e47d55510e)

**Comment by jacobpennington** [comment]
@msj220001 You can get a more detailed answer in the [paper here](https://www.nature.com/articles/s41592-024-02232-7#Sec10) in the "Template deconvolution" section of the methods.

The short answer is no, those parameters are not directly related to threshold crossings in the way you're depicting. In the spike detection step, very roughly speaking, `Th_universal` sets how closely the preprocessed and drift-corrected waveforms at a given timepoint have to match one of the "universal" spike templates to detect a spike at that time point (determined via dot product between template and data). After those spikes are detected, some clustering and other steps are done to get a set of "learned" templates that better represent the data, and the detection step is repeated using those templates.

**Comment by msj220001** [comment]
> @msj220001 You can get a more detailed answer in the [paper here](https://www.nature.com/articles/s41592-024-02232-7#Sec10) in the "Template deconvolution" section of the methods.
> 
> The short answer is no, those parameters are not directly related to threshold crossings in the way you're depicting. In the spike detection step, very roughly speaking, `Th_universal` sets how closely the preprocessed and drift-corrected waveforms at a given timepoint have to match one of the "universal" spike templates to detect a spike at that time point (determined via dot product between template and data). After those spikes are detected, some clustering and other steps are done to get a set of "learned" templates that better represent the data, and the detection step is repeated using those templates.

Thanks! The paper was helpful.
